marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/ExecEndpointImpl.java

```diff
@@ -166,32 +171,78 @@ private boolean processOutput(CallContextImpl callContext){
-                                    callContext.getEndpoint().getEndpointPath(), getCallCount(), callContext.getWorkUnit());
-                            return false;
+                if (throwable != null) {
+                    if (getErrorListener() == null) {
+                        logger.error("Error while calling " + getEndpoint().getEndpointPath(), throwable);
+                        throw new RuntimeException("Error while calling " + getEndpoint().getEndpointPath(), throwable);
+                    } else {
+                        try {
+                            if (retryCount < DEFAULT_MAX_RETRIES - 1) {
+                                error = getErrorListener().processError(retryCount, throwable, callContext);
+                            } else {
+                                error = ErrorDisposition.SKIP_CALL; //used to be STOP_ALL_CALLS
+                            }
+                        } catch (Throwable throwable1) {
+                            logger.error("Error Listener failed with " , throwable1);
+                            error = ErrorDisposition.STOP_ALL_CALLS;
+                        }
+
+                        switch (error) {
+                            case RETRY:
+                                continue;
+
+                            case SKIP_CALL:
+                                if(callContext.getEndpoint().allowsEndpointState()) {
+                                    callContext.withEndpointState((InputStream) null);
+                                }
+                                return output;
+
+                            case STOP_ALL_CALLS:
+                                getCallerThreadPoolExecutor().shutdown();
                         }
-                        return true;
-                    default:
-                        throw new MarkLogicInternalException(
-                                "unexpected state for "+callContext.getEndpoint().getEndpointPath()+" during loop: "+getPhase().name()
-                        );
+                    }
                 }
+            }
+            return output;
+        }
+
+        private void processOutput() {
+            CallContextImpl callContext = getCallContext();
+            if(callContext != null) {
+                while (processOutput(callContext));
+            }
+        }
+
+        private boolean processOutput(CallContextImpl callContext){
+            InputStream output = processExec(callContext);
+
+            switch(getPhase()) {
+                case INTERRUPTING:
+                    setPhase(WorkPhase.INTERRUPTED);
+                    logger.info("exec interrupted endpoint={} count={} work={}",
+                            callContext.getEndpoint().getEndpointPath(), getCallCount(), callContext.getWorkUnit());
+                    return false;
+                case RUNNING:
+                    if (output == null ) {
```

Comment:
Would it be better to check `aliveCallContextCount.get() == 0` instead of `getCallerThreadPoolExecutor().getActiveCount() <= 1` here as well?

Reply:
Yes. Also changed in OutputEndpointImpl as well.